File: app/src/main/java/net/ktnx/mobileledger/json/v1_32/TransactionListParser.java
/*
 * Copyright © 2020 Damyan Ivanov.
 * This file is part of MoLe.
 * MoLe is free software: you can distribute it and/or modify it
 * under the term of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your opinion), any later version.
 *
 * MoLe is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License terms for details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MoLe. If not, see <https://www.gnu.org/licenses/>.
 */

package net.ktnx.mobileledger.json.v1_32;

import com.fasterxml.jackson.databind.MappingIterator;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectReader;

import net.ktnx.mobileledger.model.LedgerTransaction;

import java.io.IOException;
import java.io.InputStream;
import java.text.ParseException;

public class TransactionListParser extends net.ktnx.mobileledger.json.TransactionListParser {

    private final MappingIterator<ParsedLedgerTransaction> iterator;

    public TransactionListParser(InputStream input) throws IOException {

        ObjectMapper mapper = new ObjectMapper();
        ObjectReader reader = mapper.readerFor(ParsedLedgerTransaction.class);
        iterator = reader.readValues(input);
    }
    public LedgerTransaction nextTransaction() throws ParseException {
        return iterator.hasNext() ? iterator.next()
                                            .asLedgerTransaction() : null;
    }
}

File: app/src/main/kotlin/net/ktnx/mobileledger/json/v1_23/TransactionListParser.kt
/*
 * Copyright © 2020 Damyan Ivanov.
 * This file is part of MoLe.
 * MoLe is free software: you can distribute it and/or modify it
 * under the term of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your opinion), any later version.
 *
 * MoLe is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License terms for details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MoLe. If not, see <https://www.gnu.org/licenses/>.
 */

package net.ktnx.mobileledger.json.v1_23

import com.fasterxml.jackson.databind.MappingIterator
import com.fasterxml.jackson.databind.ObjectMapper
import java.io.InputStream
import java.text.ParseException
import net.ktnx.mobileledger.json.TransactionListParser as BaseParser
import net.ktnx.mobileledger.model.LedgerTransaction

class TransactionListParser(input: InputStream) : BaseParser() {
    private val iterator: MappingIterator<ParsedLedgerTransaction>

    init {
        val mapper = ObjectMapper()
        val reader = mapper.readerFor(ParsedLedgerTransaction::class.java)
        iterator = reader.readValues(input)
    }

    @Throws(ParseException::class)
    override fun nextTransaction(): LedgerTransaction? =
        if (iterator.hasNext()) iterator.next().asLedgerTransaction() else null
}
